drivers/gpio/gpio_mcp23s17.c

```diff
@@ -0,0 +1,478 @@
+ */
+static int write_port_regs(struct device *dev, u8_t reg, u16_t value)
+{
+	struct mcp23s17_drv_data *const drv_data =
+		(struct mcp23s17_drv_data *const)dev->driver_data;
+	int ret;
+	u16_t port_data;
+
+	LOG_DBG("MCP23S17: Write: REG[0x%X] = 0x%X, REG[0x%X] = 0x%X",
+		reg, (value & 0xFF), (reg + 1), (value >> 8));
+
+	port_data = sys_cpu_to_le16(value);
+
+	u8_t addr = MCP23S17_ADDR;
+	u8_t buffer_tx[2] = { addr, reg };
+
+	const struct spi_buf tx_buf[2] = {
+		{
+			.buf = buffer_tx,
+			.len = 2,
+		},
+		{
+			.buf = (u8_t *)&port_data,
+			.len = 2,
+		}
+	};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = 2,
+	};
+
+	ret = spi_write(drv_data->spi, &drv_data->spi_cfg, &tx);
+	if (ret) {
+		LOG_DBG("spi_write FAIL %d\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Setup the pin direction (input or output)
+ *
+ * @param dev Device struct of the MCP23S17
+ * @param pin The pin number
+ * @param flags Flags of pin or port
+ *
+ * @return 0 if successful, failed otherwise
+ */
+static int setup_pin_dir(struct device *dev, u32_t pin, int flags)
+{
+	struct mcp23s17_drv_data *const drv_data =
+		(struct mcp23s17_drv_data *const)dev->driver_data;
+	u16_t *dir = &drv_data->reg_cache.iodir;
+	u16_t *output = &drv_data->reg_cache.gpio;
+	int ret;
+
+	if ((flags & GPIO_OUTPUT) != 0U) {
+		if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0U) {
+			*output |= BIT(pin);
+		} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0U) {
+			*output &= ~BIT(pin);
+		}
+		*dir &= ~BIT(pin);
+	} else {
+		*dir |= BIT(pin);
+	}
+
+	ret = write_port_regs(dev, REG_GPIO_PORTA, *output);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = write_port_regs(dev, REG_IODIR_PORTA, *dir);
+
+	return ret;
+}
+
+/**
+ * @brief Setup the pin pull up/pull down status
+ *
+ * @param dev Device struct of the MCP23S17
+ * @param pin The pin number
+ * @param flags Flags of pin or port
+ *
+ * @return 0 if successful, failed otherwise
+ */
+static int setup_pin_pullupdown(struct device *dev, u32_t pin, int flags)
+{
+	struct mcp23s17_drv_data *const drv_data =
+		(struct mcp23s17_drv_data *const)dev->driver_data;
+	u16_t port;
+	int ret;
+
+	/* Setup pin pull up or pull down */
+	port = drv_data->reg_cache.gppu;
+
+	/* pull down == 0, pull up == 1 */
+	if ((flags & GPIO_PULL_DOWN) != 0U) {
+		return -ENOTSUP;
+	}
+
+	WRITE_BIT(port, pin, (flags & GPIO_PULL_UP) != 0U);
+
+	ret = write_port_regs(dev, REG_GPPU_PORTA, port);
+	if (ret == 0) {
+		drv_data->reg_cache.gppu = port;
+	}
+
+	return ret;
+}
+
+static int mcp23s17_config(struct device *dev,
+			   gpio_pin_t pin, gpio_flags_t flags)
+{
+	struct mcp23s17_drv_data *const drv_data =
+		(struct mcp23s17_drv_data *const)dev->driver_data;
+	int ret;
+
+	/* Can't do SPI bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	if ((flags & GPIO_OPEN_DRAIN) != 0U) {
+		ret = -ENOTSUP;
+		goto done;
+	};
```

Comment:
@mnkp good test `gpio_api_1pin` I was missing the devices doesn't support open drain